### PR TITLE
Implement labels with async-profiler's contextId

### PR DESCRIPTION
**Pull request by korniltsev**
The api is:
```
Pyroscope.LabelsWrapper.run(new LabelsSet("thread_name", Thread.currentThread().getName()), () -> {
    // ...
});
```

See https://github.com/jvm-profiling-tools/async-profiler/pull/576 for reference

### CODE REVIEW

File: agent/src/main/java/io/pyroscope/api/Tags.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+public class Tags {
```
Comment (petethepig) [inline agent/src/main/java/io/pyroscope/api/Tags.java:16]: Similar to my comment in the other PR (https://github.com/pyroscope-io/pyroscope/pull/1096/files#r872029411), we should call these labels for the sake of consistency with other profilers.

File: agent/src/main/java/io/pyroscope/api/Tags.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                        }
+                    }
+                }
+            }
+            context.set(prev);
+            contextId.set(prevContextId);
+            AsyncProfiler.getInstance().setContextId(prevContextId);
+        }
+    }
+
```
Comment (petethepig) [inline agent/src/main/java/io/pyroscope/api/Tags.java:257]: extra new lines?

File: agent/src/main/java/io/pyroscope/api/Tags.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+                '}';
+        }
+    }
+
+    public static class Scope implements AutoCloseable {
+        final HashMap<Long, Long> prev;
+        final Long prevContextId;
+
+        public Scope(Object... args) { //todo create constructor for 2, 4, 6 args
+            if (args.length %2  != 0) {
```
Comment (petethepig) [inline agent/src/main/java/io/pyroscope/api/Tags.java:186]:
```suggestion
            if (args.length %2 != 0) {
```

File: agent/src/main/java/io/pyroscope/api/Tags.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                throw new IllegalArgumentException();
```
Comment (petethepig) [inline agent/src/main/java/io/pyroscope/api/Tags.java:187]: is it common to provide an error message in this case? (I'm a Java noob)

Comment (korniltsev) [reply agent/src/main/java/io/pyroscope/api/Tags.java]: I added error message

File: agent/src/main/java/io/pyroscope/api/Tags.java (hunk 3 above)
Comment (korniltsev) [reply agent/src/main/java/io/pyroscope/api/Tags.java]: I reformatted code by hand

File: agent/src/main/java/io/pyroscope/api/Tags.java (hunk 2 above)
Comment (korniltsev) [reply agent/src/main/java/io/pyroscope/api/Tags.java]: removed

File: agent/src/main/java/io/pyroscope/api/Tags.java (hunk 1 above)
Comment (korniltsev) [reply agent/src/main/java/io/pyroscope/api/Tags.java]: Renamed to 'Labels' and 'LabelsSet'

Comment (korniltsev) [reply agent/src/main/java/io/pyroscope/api/Tags.java]: Please let me know if you like new names

**Comment by korniltsev** [comment]
Hi @petethepig 
I've addressed your comments.
I've also added code for uploading, demo app, small unit test, so you may want to take a look again.
Also this is still WIP, I am still working on implementing more concurrent contexts.

**Review by petethepig — APPROVED** [review]
LGTM!

**Pull request merged**